File: backend/src/domain/user/StaffBuilder.ts
import {Staff} from "./Staff";
import {Clinic} from "@/domain/clinic";
import {StaffRole} from "./StaffRole";

export class StaffBuilder {
  private id?: string
  private firstname?: string
  private lastname?: string
  private patronymic?: string
  private birthday?: Date
  private avatarUrl?: string
  private phone?: string
  private email?: string
  private clinic?: Clinic
  private role?: StaffRole

  public setClinic(clinic: Clinic) {
    this.clinic = clinic
    return this;
  }

  public setRole(role: StaffRole) {
    this.role = role
    return this;
  }

  public setId(id: string) {
    this.id = id
    return this;
  }

  public setFirstname(firstname: string) {
    this.firstname = firstname
    return this;
  }

  public setLastname(lastname: string) {
    this.lastname = lastname
    return this;
  }

  public setPatronymic(patronymic: string) {
    this.patronymic = patronymic
    return this;
  }

  public setBirthday(birthday: Date) {
    this.birthday = birthday
    return this;
  }

  public setAvatarUrl(avatarUrl: string) {
    this.avatarUrl = avatarUrl
    return this;
  }

  public setPhone(phone: string) {
    this.phone = phone
    return this;
  }

  public setEmail(email: string) {
    this.email = email
    return this;
  }

  public build() {
    return new Staff(
      this.id,
      this.firstname,
      this.lastname,
      this.patronymic,
      this.birthday,
      this.avatarUrl,
      this.phone,
      this.email,
      this.clinic,
      this.role,
    )
  }
}

File: backend/src/main/java/com/hypocrates/hypocrates/domain/StaffBuilder.java
package com.hypocrates.hypocrates.domain;


import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.Date;

@Accessors(chain = true) @Setter
public class StaffBuilder {
    private Long id;
    private String firstname;
    private String lastname;
    private String patronymic;
    private Date birthday;
    private String avatarUrl;
    private String email;
    private String phone;
    private Boolean emailIsActive;
    private Boolean phoneIsActive;
    private Clinic clinic;
    private StaffRole role;
    private String password;

    public StaffBuilder setUser(User user) {
        this.id = user.getId();
        this.firstname = user.getFirstname();
        this.lastname = user.getLastname();
        this.patronymic = user.getPatronymic();
        this.birthday = user.getBirthday();
        this.avatarUrl = user.getAvatarUrl();
        this.email = user.getEmail();
        this.phone = user.getPhone();
        this.emailIsActive = user.getEmailIsActive();
        this.phoneIsActive = user.getPhoneIsActive();

        this.password = user.getPassword();
        return this;
    }

    public Staff createStaff() {
        return new Staff(id, firstname, lastname, patronymic, birthday, avatarUrl, email, phone, emailIsActive, phoneIsActive, clinic, role, password);
    }
}
